source/extensions/filters/http/rbac/rbac_filter.cc

```diff
@@ -77,24 +77,46 @@ Http::FilterHeadersStatus RoleBasedAccessControlFilter::decodeHeaders(Http::Head
                 callbacks_->connection()->ssl()->subjectPeerCertificate()
           : "none",
       headers, callbacks_->requestInfo().dynamicMetadata().DebugString());
+  std::string effective_policy_id;
   const absl::optional<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl>& shadow_engine =
       config_->engine(callbacks_->route(), EnforcementMode::Shadow);
   if (shadow_engine.has_value()) {
+    std::string shadow_resp_code = "200";
     if (shadow_engine->allowed(*callbacks_->connection(), headers,
-                               callbacks_->requestInfo().dynamicMetadata())) {
+                               callbacks_->requestInfo().dynamicMetadata(), effective_policy_id)) {
       ENVOY_LOG(debug, "shadow allowed");
       config_->stats().shadow_allowed_.inc();
     } else {
       ENVOY_LOG(debug, "shadow denied");
       config_->stats().shadow_denied_.inc();
+      shadow_resp_code = "403";
+    }
+
+    const auto& filter_metadata = callbacks_->requestInfo().dynamicMetadata().filter_metadata();
+    const auto filter_it = filter_metadata.find(HttpFilterNames::get().Rbac);
+    if (filter_it != filter_metadata.end()) {
+      ProtobufWkt::Struct metrics;
+
+      if (effective_policy_id != "") {
+        ProtobufWkt::Value policy_id;
+        policy_id.set_string_value(effective_policy_id);
+        (*metrics.mutable_fields())["shadow_effective_policyID"] = policy_id;
+      }
+
+      ProtobufWkt::Value resp_code;
+      resp_code.set_string_value(shadow_resp_code);
+      (*metrics.mutable_fields())["shadow_response_code"] = resp_code;
+
+      auto filter_meta = filter_metadata.at(HttpFilterNames::get().Rbac);
+      filter_meta.MergeFrom(metrics);
     }
   }
 
   const absl::optional<Filters::Common::RBAC::RoleBasedAccessControlEngineImpl>& engine =
       config_->engine(callbacks_->route(), EnforcementMode::Enforced);
   if (engine.has_value()) {
     if (engine->allowed(*callbacks_->connection(), headers,
-                        callbacks_->requestInfo().dynamicMetadata())) {
+                        callbacks_->requestInfo().dynamicMetadata(), effective_policy_id)) {
```

Comment:
Why don't we use this  value here? Can we avoid cost by passing nullptr and not filling/copying?

Reply:
we only need effective_policy_id for shadow case for now(can add later if needed); nullptr is invalid since ```allowed``` takes ```std::string&``` ?

Reply:
I would make it take `std::string*` then.

Reply:
updated.